### PR TITLE
Improve consistency with other JS SDKs

### PR DESCRIPTION
**Pull request by kearfy**
This PR:
- Improves typescript types
- Make the behaviour and argument list of functions more consistent with other JS SDKs

### CODE REVIEW

File: src/app/mod.rs [hunk 1]
```diff
@@ -335,16 +336,15 @@ impl Surreal {
 			},
 		};
 		let num_statements = response.num_statements();
-		let response = if num_statements > 1 {
+		let response = {
 			let mut output = Vec::<Value>::with_capacity(num_statements);
 			for index in 0..num_statements {
-				output.push(response.take(index)?);
+				let v: Value = response.take(index)?;
```
Comment (rushmorem) [inline src/app/mod.rs:342]:
```suggestion
				let v = response.take(index)?;
```

The type hint shouldn't be necessary here. They compiler should be able to infer it from `Vec<Value>`.

File: src/app/mod.rs [hunk 2]
```diff
@@ -359,14 +368,15 @@ impl Surreal {
 	/// // Select a specific record from a table
 	/// const person = await db.select('person:h5wxrf2ewk8xjxosxtyc');
 	/// ```
-	pub async fn select(&self, resource: String) -> Result<JsValue, Error> {
+	#[wasm_bindgen]
```
Comment (rushmorem) [inline src/app/mod.rs:371]: Why is this necessary in this case? The other methods don't have it.

Comment (kearfy) [reply src/app/mod.rs]: It's not, thanks for spotting

File: src/app/mod.rs [hunk 3]
```diff
@@ -326,25 +332,28 @@ impl Surreal {
...
 			for index in 0..num_statements {
 				output.push(response.take(index)?);
 			}
-			Value::from(output)
-		} else {
-			response.take(0)?
+			Value::from(Array::from(output))
 		};
-		Ok(to_value(&response.into_json())?)
+		Ok(TsArrayUnknown::from_value(response)?)
```
Comment (rushmorem) [inline src/app/mod.rs:356]:
```suggestion
		TsArrayUnknown::from_value(response)
```

File: src/app/mod.rs [hunk 4]
```diff
@@ -359,14 +368,15 @@ impl Surreal {
...
+	#[wasm_bindgen]
+	pub async fn select(&self, resource: String) -> Result<TsArrayRecordUnknown, Error> {
 		let response = match resource.parse::<Range>() {
 			Ok(range) => {
 				self.db.select(Resource::from(range.tb)).range((range.beg, range.end)).await?
 			}
 			Err(_) => self.db.select(Resource::from(resource)).await?,
 		};
-		Ok(to_value(&response.into_json())?)
+		Ok(TsArrayRecordUnknown::from_value(response)?)
```
Comment (rushmorem) [inline src/app/mod.rs:378]:
```suggestion
		TsArrayRecordUnknown::from_value(response)
```

File: src/app/mod.rs [hunk 5]
```diff
@@ -384,16 +394,21 @@ impl Surreal {
...
 		let response = match data.is_undefined() {
 			true => self.db.create(resource).await?,
 			false => {
 				let data = json(&from_value::<Json>(data)?.to_string())?;
 				self.db.create(resource).content(data).await?
-			},
+			}
 		};
-		Ok(to_value(&response.into_json())?)
+		Ok(TsArrayRecordUnknown::from_value(response)?)
```
Comment (rushmorem) [inline src/app/mod.rs:410]:
```suggestion
		TsArrayRecordUnknown::from_value(response)
```

File: src/app/mod.rs [hunk 6]
```diff
@@ -436,9 +456,9 @@ impl Surreal {
 			false => {
 				let data = json(&from_value::<Json>(data)?.to_string())?;
 				update.content(data).await?
-			},
+			}
 		};
-		Ok(to_value(&response.into_json())?)
+		Ok(TsArrayRecordUnknown::from_value(response)?)
```
Comment (rushmorem) [inline src/app/mod.rs:460]:
```suggestion
		TsArrayRecordUnknown::from_value(response)
```

File: src/app/mod.rs [hunk 7]
```diff
@@ -459,14 +479,19 @@ impl Surreal {
...
+	) -> Result<TsArrayRecordUnknown, Error> {
+		let data = JsValue::from(data);
 		let update = match resource.parse::<Range>() {
 			Ok(range) => self.db.update(Resource::from(range.tb)).range((range.beg, range.end)),
 			Err(_) => self.db.update(Resource::from(resource)),
 		};
 		let data = json(&from_value::<Json>(data)?.to_string())?;
 		let response = update.merge(data).await?;
-		Ok(to_value(&response.into_json())?)
+		Ok(TsArrayRecordUnknown::from_value(response)?)
```
Comment (rushmorem) [inline src/app/mod.rs]:
```suggestion
		TsArrayRecordUnknown::from_value(response)
```

File: src/app/opt/patch.rs [hunk 8]
```diff
@@ -22,11 +14,10 @@ pub enum Patch {
 	},
 	Replace {
 		path: String,
-		value: Value,
+		value: String,
```
Comment (rushmorem) [inline src/app/opt/patch.rs:17]: Are you sure about this? Have you tried it?

File: src/app/opt/patch.rs [hunk 9]
```diff
@@ -22,11 +14,10 @@ pub enum Patch {
...
 	},
 	Change {
 		path: String,
-		#[serde(with = "DiffDef")]
-		diff: Diff,
+		value: String,
```
Comment (rushmorem) [inline src/app/opt/patch.rs:21]: This too...

File: src/app/types.rs [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+impl TsArrayRecordUnknown {
+	pub fn from_value(value: Value) -> Result<Self, Error> {
+		let value = match value {
+			Value::Array(v) => v,
+			_ => Array::from(value),
+		};
+
+		for v in value.clone() {
+			if !matches!(v, Value::Object(_)) {
```
Comment (rushmorem) [inline src/app/types.rs:133]:
```suggestion
		for v in &value {
			if !v.is_object() {
```

File: src/app/opt/patch.rs (hunk 8 above)
Comment (kearfy) [reply src/app/opt/patch.rs:17]: This is for type safety, the string will get auto converted into a surrealql value

File: src/app/opt/patch.rs (hunk 9 above)
Comment (kearfy) [reply src/app/opt/patch.rs:21]: This should accept a string, have a look at this example from how the patch change operation works in SurrealQL:

```sql
update person:1 patch [{ op: "change", path: "/name", value: "@@ -1,4 +1,4 @@\n te\n-st\n+mp\n" }];
```

**Review by rushmorem — APPROVED** [review]
